### PR TITLE
Migrate gradle from groovy to kotlin dsl

### PR DESCRIPTION
**Pull request by RenFraser**
Hi @fwcd, this is a PR to migrate the build scripts from Groovy to Kotlin DSL. It's not a perfect implementation but think that it'll be a good start that we can build upon.

To create this change I've:

1. Run a conversion script using [GradleKotlinConverter ](https://github.com/bernaferrari/GradleKotlinConverter).
2. Fixed up any missed changes and manually went through the DSL files to ensure they worked.
3. Compared the diff against [your migration PR from 2020](https://github.com/fwcd/kotlin-language-server/pull/193) and used your changes as a reference to fix any remaining issues.
4. Ran the build on my GitHub actions with a [successful outcome](https://github.com/RenFraser/kotlin-language-server/actions/runs/4089845504)

I'd love to get this merged because I'd like to introduce some static analysis tools to the build, such as [ktlint](https://github.com/pinterest/ktlint) and [Detekt](https://github.com/detekt/detekt)

### CODE REVIEW

File: server/build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+val debugPort = 8000
+val debugArgs = "-agentlib:jdwp=transport=dt_socket,server=y,address=8000,suspend=n,quiet=y"
+
+version = projectVersion
+
+val mainClassName = "org.javacs.kt.MainKt"
+val applicationName = "kotlin-language-server"
+
+application {
+    mainClassName = mainClassName
```
Comment (themkat) [inline server/build.gradle.kts:23]: `mainClassName` seems to be shadowed here, so it isn't set to the variable defined above. This causes the language server to crash on startup on my machine. If I try to run the language server from the command line (`./server/build/install/server/bin/kotlin-language-server`) it does not seem to set up the java-command correctly, as it prints the usage helper. This seems to be due to mainClass not being set correctly, and therefore no main class name being used in the Java command. (the resulting script simply runs java without any arguments causing the usage-message).

Got everything working if used another name for the variable in line 19 🙂

**Review by themkat — COMMENTED** [review]
Nice work! 🙂  Took a quick look and tested it. Have a few comments below.

File: shared/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,22 @@
+plugins {
+    id("maven-publish")
+    kotlin("jvm")
+}
+
+repositories {
+    jcenter()
```
[the remark this answers is not in the document]
Comment (RenFraser) [reply shared/build.gradle.kts]: That sounds like a great idea. I've added this change to use `mavenCentral()`, squashed it into the original commit and push the branch for this PR.

File: server/build.gradle.kts (hunk 1 above)
Comment (RenFraser) [reply server/build.gradle.kts]: Ok, thank you for helping with out on the testing front @themkat! I've renamed it on line 19 and renamed the references under the `application` block and `debugRun` + `debugStartScripts` tasks and squashed the change into the original commit.

**Comment by themkat** [comment]
Tried building for both Java 11 and 17 locally, and both succeed on my machine 😱 Unsure on why the CI fails here. Could it be that the tests that fails are a bit flaky? Will try to have a deeper look at that...

**Comment by RenFraser** [comment]
> tests that fails are a bit flaky

Ok, I've bumped the kotlin version in the top level `build.gradle.kts` to match the newly committed version of 1.8.10. If you can approve the workflow and it passes then it could be just that there was a version mismatch that needed to be updated?

**Comment by themkat** [comment]
> > tests that fails are a bit flaky
> 
> Ok, I've bumped the kotlin version in the top level `build.gradle.kts` to match the newly committed version of 1.8.10. If you can approve the workflow and it passes then it could be just that there was a version mismatch that needed to be updated?

Nice! Still waiting for accesses, but if I get them I will approve the workflow 🙂

**Comment by fwcd** [comment]
@themkat I've sent out an invite. Did you get my e-mail?

**Comment by themkat** [comment]
> @themkat I've sent out an invite. Did you get my e-mail?

Yes 😄 Thank you! Will try to have a look later tonight. Sadly I need some time to relax after a long day at work first 🙁

**Review by themkat — COMMENTED** [review]
Finally managed to have time for another look 🙂 Some minor things. As I was able to edit the gradle Kotlin DSL scripts using the language server itself (almost) without any issues, I suggest we create some issues for my comments below. Our show-stopper for introducing Kotlin DSL in Gradle files before was that the files were almost only errors when editing with the language server 😱 

Great work! Will let you have a chance to give your replies, if any, on the below comments. If you have none, I will approve and merge this tomorrow 🙂

File: server/build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+    classpath(sourceSets.main.get().runtimeClasspath)
+    standardInput = System.`in`
+
+    jvmArgs(debugArgs)
+    doLast {
+        println("Using debug port $debugPort")
+    }
+}
+
+tasks.register<CreateStartScripts>("debugStartScripts") {
```
[the remark this answers is not in the document]
Comment (RenFraser) [reply server/build.gradle.kts:112]: I'm using IntelliJ :P I get a few deprecated warnings but no big errors like that. Here's my view of the same block.

image

File: build.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,35 @@
+// TODO: change the hardcoded 1.6.10 to kotlinVeresion that lives in project props
```
[the remark this answers is not in the document]
Comment (RenFraser) [reply build.gradle.kts:1]: Good idea! I've opened #419 to address this.

File: server/build.gradle.kts [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+tasks.register<Sync>("installDebugDist") {
+    dependsOn("installDist")
+    finalizedBy("debugStartScripts")
+}
+
+tasks.getByName("processTestResources") {
+    dependsOn("copyPropertiesToTestWorkspace", "copyPropertiesToDSLTestWorkspace")
+}
+
+tasks.withType<Test>() {
```
[the remark this answers is not in the document]
Comment (RenFraser) [reply server/build.gradle.kts:129]: It looks like it's the conditions as the above. Some sort of elusive error in the language server. I've created  #420 to address it :)

File: server/build.gradle.kts (hunk 3 above)
[the remark this answers is not in the document]
Comment (themkat) [reply server/build.gradle.kts:112]: Oh. Then you are using IntelliJs completion and error checking mechanisms. Unless it support custom language servers now?

**Review by themkat — APPROVED** [review]
Thanks for making the issues! Will merge this PR now. Sorry for taking some time!

File: build.gradle.kts [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    apply(plugin = "maven-publish")
+
+    publishing {
+        repositories {
+            maven {
+                name = "GitHubPackages"
+                url = uri("https://maven.pkg.github.com/fwcd/kotlin-language-server")
+                credentials {
+                    username = project.findProperty("gpr.user") as String? ?: System.getenv("GPR_USERNAME")
+                    password = project.findProperty("gpr.key") as String? ?: System.getenv("GPR_PASSWORD")
```
Comment (fwcd) [inline build.gradle.kts:24]: These env vars changed, which tripped up our release workflow (though arguably the new names are less CI-specific).

**Pull request merged**